core/src/main/java/hudson/ClassicPluginStrategy.java

```diff
@@ -250,14 +250,18 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
                 createClassLoader(paths, dependencyLoader, atts), disableFile, dependencies, optionalDependencies);
     }
 
-    private static void fix(Attributes atts, List<PluginWrapper.Dependency> optionalDependencies) {
+    private void fix(Attributes atts, List<PluginWrapper.Dependency> optionalDependencies) {
         String pluginName = atts.getValue("Short-Name");
         
         String jenkinsVersion = atts.getValue("Jenkins-Version");
         if (jenkinsVersion==null)
             jenkinsVersion = atts.getValue("Hudson-Version");
-        
-        optionalDependencies.addAll(DetachedPluginsUtil.getImpliedDependencies(pluginName, jenkinsVersion));
+
+        for (Dependency d : DetachedPluginsUtil.getImpliedDependencies(pluginName, jenkinsVersion)) {
```

Comment:
What will happen if a newer version of plugin should be loaded by a common mechanism later? My code knowledge got a bit rusty

Reply:
Do not think I understand the question. What “common mechanism”? Newer version of which plugin—`pluginName` or `d.shortName`?

As seen [here](https://github.com/jenkinsci/jenkins/pull/4000/files#diff-c06f61bb7c43d48aa0554204a18832e6R605), the only behavioral change comes when there is an implied dependency on a detached plugin from something which was already in `plugins/*.jpi` _and_ that detached plugin is not already installed (really missing, not just disabled). That is the scenario you see in things like Docker packaging, where plugins are dropped into the `plugins` directory by some arbitrary tool; when you are using the Jenkins GUI (or I guess CLI) to manage plugins, the plugin manager would have added the detached dependency automatically already (in the latest version on the update center).

After this point, if there are any upgrades on another Jenkins startup, nothing is different than the usual case.

Reply:
IIUC, a different plugin would fail – right now, the existing plugin fails to load but the dependency would later be installed to support another plugin, in the future the newly added plugin with a (then outdated) dependency would fail.

Still, this seems like a net improvement.

Reply:
> right now, the existing plugin fails to load

No, it _does_ load, because the implied dependencies on detached plugins are marked optional. So it loads, and then _probably_ it works.

> in the future the newly added plugin with a (then outdated) dependency would fail.

Sorry, I did not follow this part at all.